activerecord/lib/active_record/attribute_methods/query.rb

```diff
@@ -18,7 +18,7 @@ def query_attribute(attr_name)
         else
           column = self.class.columns_hash[attr_name]
           if column.nil?
```

Comment:
This is pretty old code (for 10 years ago https://github.com/rails/rails/commit/5b801b596014255ecf55dcf58c82cbf061faa08b#diff-dfe09551f2188457e7e1891d86047c57R231), since we now have Attribute API and the responsibility of type casting was extracted from columns to Type objects.

If an attribute has a type defined on the model by like `attribute :attr_name, :json`, a value from `read_attribute` (`self[attr_name]`) is considered as casted.

https://github.com/rails/rails/blob/16ebcde31f5f31ecb94f233ffcc195842a7c3ca5/activerecord/lib/active_record/attribute_methods/query.rb#L12-L13

So can you change the line to check whether or not an attribute has a type?

```suggestion
          if !type_for_attribute(attr_name) { false }
```

Reply:
I changed the line, but I'm not sure the code was fixed as you intended.